src/include/index/ivf_pq_index.h

```diff
@@ -0,0 +1,1576 @@
+  using pq_code_type = uint8_t;
+
+  using pq_vector_feature_type = pq_code_type;
+
+  // The pq_centroids store the (unencoded) centroids for each subspace
+  using flat_vector_feature_type = score_type;
+
+ private:
+  using flat_storage_type = ColMajorMatrix<pq_code_type>;
+  using tdb_flat_storage_type = ColMajorMatrix<pq_code_type>;
+
+  using pq_storage_type = ColMajorPartitionedMatrix<  // was: storage_type
+      pq_code_type,                                   // was: feature_type
+      partitioned_ids_type,
+      indices_type>;
+
+  using tdb_pq_storage_type =
+      tdbColMajorPartitionedMatrix<  // was: tdb storage_type
+          pq_code_type,              // was: feature_type
+          partitioned_ids_type,
+          indices_type>;
+
+  /*
+   * We need to store three different sets of centroids.
+   *   - The flat ivf centroids_ which are uncompressed and used to partition
+   *     the entire training set with train / add+compress patterns.
+   *   - The pq ivf centroids - compressed version of the ivf centroids.  These
+   *     are used in queries with symmetric distance.  They are also used in
+   *     the compress / train / add pattern.
+   *   - The cluster_centroids_ that partition each subspace of the training set
+   *     into another set of partitions. There are num_clusters of these, with
+   *     each centroid being of size dimension_.  These are used to build the
+   *     distance_tables_ and compress the training set and ivf centroids
+   */
+  using flat_ivf_centroid_storage_type =
+      ColMajorMatrix<flat_vector_feature_type>;
+  using tdb_flat_ivf_centroid_storage_type =
+      tdbColMajorMatrix<flat_vector_feature_type>;
+
+  using pq_ivf_centroid_storage_type = ColMajorMatrix<pq_vector_feature_type>;
+  using tdb_pq_ivf_centroid_storage_type =
+      tdbColMajorMatrix<pq_vector_feature_type>;
+
+  using cluster_centroid_storage_type =
+      ColMajorMatrix<flat_vector_feature_type>;
+  using tdb_cluster_centroid_storage_type =
+      tdbColMajorMatrix<flat_vector_feature_type>;
+
+  constexpr static const IndexKind index_kind_ =
+      IndexKind::IVFPQ;  // was: IVFFlat
+
+  /****************************************************************************
+   * Index group information
+   ****************************************************************************/
+
+  /** The timestamp at which the index was created */
+  uint64_t timestamp_{0};
+
+  std::unique_ptr<ivf_pq_group<ivf_pq_index>> group_;
+
+  /****************************************************************************
+   * Index representation
+   ****************************************************************************/
+
+  // Cached information about the partitioned vectors in the index
+  uint64_t dimension_{0};
+  uint64_t num_partitions_{0};
+
+  // Cached information about the pq encoding
+  uint64_t num_subspaces_{0};
+  uint64_t sub_dimension_{0};
+  constexpr static const uint64_t bits_per_subspace_{8};
```

Comment:
Shouldn't `bits_per_subspace_` or `num_clusters_` be a configurable parameter?

Reply:
In theory yes.  But.  Since the pq-encoded vector stores the subspace indices contiguously, we essentially have the options of 8 bits or 16 bits.  Otherwise we have to pick out the right set of bits from contiguous bytes and turn that into something suitable for indexing into the centroids.  That would require some bit twiddling and so forth.  Performance is better if one just uses 8 (it seems like a lot of systems do that) -- and accuracy does not seem horrible if IVF has reasonable partitions.  Alot f systems I looked at just use 8 -- but I suppose we could offer a few different values.  We would want to pick values that are easy to bit twiddle.  We might also be able to use SIMD to do multiple lookups at once, which might help amortize the bit twiddling cost.  Though for those kinds of operations we should probably be using SIMD instructions even for 8-bits.

Both of these should be new PRs:
* configurable `bits_per_subspace`
* use SIMD instructions to reconstruct vectors
*The latter is actually very important for performance* (unless somehow the compiler would do it, which I doubt)

Reply:
I agree for splitting this work out to new PRs. Lets add some TODOs to make sure we don't forget this.

Reply:
I created two new shortcut stories for this (45148 and 45149).